### PR TITLE
manager: refuse manual start and stop of device units

### PR DESCRIPTION
**Pull request by jsynacek**
Manually starting and stopping device units that don't exist
results in a confusing behavior.

### CODE REVIEW

**Comment by evverx** [comment]
I've seen such synchronization points in bash scripts:

``` sh
...
# waiting for some.device appearance
while ! systemctl start some.device; do :; done
```

Does this PR break all those scripts?

**Comment by arvidjaar** [comment]
@evverx 
Really, that's what `is-active` is for.

**Comment by evverx** [comment]
@arvidjaar, I think people should use `RUN{program}+=` or/and `SYSTEMD_WANTS+=...` in the udev rules.

But some people run `systemctl start some.device` in bash scripts. Why not? Why should we break this?

**Comment by arvidjaar** [comment]
> Why should we break this?

I think it depends on whether this patch fixes real problem (and how serious it is) or it is just bad feeling.  So far commit message looks more like bad feeling.

@jsynacek Could you explain the reason for this patch?

**Comment by jsynacek** [comment]
> But some people run systemctl start some.device in bash scripts

Some people also run the same command and don't expect it to hang and time out, because they are not developers and don't know that it's supposed to behave like that.

Also, `systemctl stop foobar.device` always finishes immediately and returns 0. I know (at least I think I know) why, but regular users don't and it's really confusing.

The same with `systemctl status foobar.device`, it always says that the unit is loaded. This one is actually quite okay, but I think it deserves some documentation.

To be honest, I think that the bash script that uses systemctl **start** to **wait** is completely crazy.

**Comment by arvidjaar** [comment]
> I think that the bash script that uses systemctl start to wait is completely crazy

Well ... comments in code suggest that it was intended this way initially. I suppose, `is-active` came later. Also do not forget that device units also may have dependencies; and so `systemctl start foo.device` will also start all dependencies. And there are even cases where it is actually useful - you must run cryptsetup to actually get your device at all; you cannot simply passively wait for it. Actually any other virtual device may potentially be configured this way (`lvm_volume.device` pulling in `lvchange -ay lvm_volume` etc).

So while using `start` as substitute for `wait` may look crazy, there are other legitimate use cases which look pretty sane.

> The same with systemctl status foobar.device, it always says that the unit is loaded.

That is unfortunate side effect of lazy loading which led us to #3856. The mere act of querying status of unit triggers attempt to load it full scale. So yes, the above is correct - unit _was_ loaded at the moment this command was executing. Of course it disappears almost instantly due to garbage collection.

May be implicit loading should be restricted to the `start` and `show` only. And I am not sure even about `show`.

**Comment by jsynacek** [comment]
> May be implicit loading should be restricted to the start and show only. And I am not sure even about show.

Restricting implicit loading only to `start` sounds like a reasonable solution. The `show` is very similar to `status` in this case.

**Comment by poettering** [comment]
This isn't right. "systemctl start /dev/sda" has always been a way to sync on /dev/sda becoming available, and I am pretty sure it should continue to be useful for that. "systemctl stop /dev/sdd" has always been a way to sync on /dev/sdd going away. And I am pretty sure we should continue to support both, as it's absolutely valid to have this in scripts. 

Note that running these commands might actually pull in other units, that may result in the device becoming available/being removed. Hence, while typically devices pop-up only due to external events (for example: user plugs in a usb stick) and hence "systemctl start" on them waits for that, in some cases they may be generated by software (think: iscsi, DM, loop, MD, …) in which case "systemctl start" might or might not pull enough in to make this happen. The net effect is the same: "do as much as you can to make this available, and if this command returns cleanly the device is there".

Yes, this might be confusing at first, but I think it makes a ton of sense to have it that way. Yes, we could document this better, and I'd love to see a patch for that, but I am very much of the opinion that the current behaviour is the right one, and scripts doing "systemctl start dev-foo.device" to sync on /dev/foo are doing things correctly and we should not break it.

systemctl status and show will cause units to be loaded as side effect, and that's definitely the right approach I am sure, simply because the mode of operation of systemd is that all units are always available, and loading is always implicit. Note that there is no unit load state of "unloaded" anyway. Because if it is unloaded systemd wouldn't have any idea of it. There's only "loaded", "masked", "not-found", "error" (plus two internal ones: "stub" and "merged"). 

Moreover, note that on recent systemd versions "systemctl status" on a unit with "not-found" or "error" will actually show a short message saying the unit doesn't exist instead of actually showing the unit data.

Long story short: I am very sure the current behaviour of "systemctl start" and of "systemctl status" is pretty OK. However, we should improve the documentation about this. Patches for that greatly welcome!

I hope this makes sense. Closing PR.
